### ISSUE TITLE
"Excessive stack depth comparing types 'FlatArray<Arr, ?>' and 'FlatArray<Arr, ?>'" false positive

### ISSUE THREAD

**Issue by DetachHead** [microsoft/TypeScript#43493]
# Bug Report

seems to be a very similar issue to #43249, where #43435 didn't seem to fix it

### 🔎 Search Terms
Excessive stack depth comparing types 'FlatArray<Arr, ?>' and 'FlatArray<Arr, ?>

### 🕗 Version & Regression Information
4.2.3-dev.20210401

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?target=6&ts=4.3.0-dev.20210401#code/CYUwxgNghgTiAEYD2A7AzgF3gMyUgXPAK4oDWKSA7igNoC6AUMulgEazwC8OeAdNtAwBZKAAcAFOxhcAfPCnwoaeADE8ASngB6LQFEAHmBBo0ASwBuCTFDCl4oURgAWiJAFtRsUygDm8DACeosbwAOQqggCCMDBQAQA80TAANPAA-DKhiijAYRFQGElxiTGpGaG84gBMAMxVAIzqDAzeGCAw2DYIakjwIPptOcpFCZgw3j5yAN4AvkA)

### 💻 Code

```ts
declare const foo: unknown[]
const bar = foo.flatMap(bar => bar as Foo) //Excessive stack depth comparing types 'FlatArray<Arr, ?>' and 'FlatArray<Arr, ?>'.(2321)

interface Foo extends Array<string> {}
```

### 🙁 Actual behavior
`Excessive stack depth comparing types 'FlatArray<Arr, ?>' and 'FlatArray<Arr, ?>` error

### 🙂 Expected behavior
no error

**Comment by ahejlsberg** [comment]
I did some investigating on this issue.

An `Excessive stack depth comparing types xxx and yyy` error is generally a red flag that our recursion depth limiter isn't working. The logic in `recursiveTypeRelatedTo` and `isDeeplyNestedType` is supposed to identify deep and possibly infinite repeating patterns of type instantiations and cut off after five levels of nesting. Only when that fails do we hit the "panic" limit of 100 recursive invocations and trigger the excessive stack depth error.

In the particular example of relating `FlatArray` instantiations (for purposes of determining variance as revealed by the odd `?` types) we fail to recognize a deeply nested type on the source side because `getRecursionIdentity` doesn't produce an identity when invoked for a type parameter. Having thought about it some, it doesn't really make sense of `getRecursionIdentity` to ever return undefined--rather, we should simply return the type instance itself. That's actually how we treat it in type inference (because we default to the type itself when `getRecursionIdentity` returns `undefined`). I'm going to put up a PR to that effect.

I may also put up a PR with a better implementation of `FlatArray`. It's not critical, but the current implementation definitely has several issues.
